# Технологии ИИ · Занятие 3 · Практика и домашнее задание

**Фамилия, имя:** _(Саргсян Ваагн)_ · **Группа:** ДИБ 401 · **Срок:** до следующей пары

Файл `auth_log.csv` должен лежать в той же папке, что и этот блокнот. Ячейки
запускаются по порядку: курсор в ячейку → **Shift + Enter**. Под каждой ячейкой
есть вопрос, ответ пишите словами в ячейку «Ответ» (два щелчка по ней —
редактирование, Shift + Enter — сохранить).

Три части:

| Часть | Что | Откуда |
|---|---|---|
| **А** | Журнал входов как таблица: запустить, прочитать, ответить | практика занятий 2–3 |
| **Б** | Пять задач ИБ → тип задачи машинного обучения | теория занятия 3 |
| **В** | Домашняя практика: изменить код под новый вопрос | новое |

Как сдавать — в самом конце блокнота.

## Если библиотек ещё нет

1. VS Code → Select Kernel (справа вверху) → Python Environments → ваш Python 3.11+.
   Предложит установить `ipykernel` — соглашайтесь.
2. Выполните ячейку ниже. Если интернета нет, а рядом с блокнотом лежит папка
   `wheels`, замените команду на `%pip install --no-index --find-links=wheels pandas scikit-learn`.
3. Подробная инструкция с картинками ошибок — `ИНСТРУКЦИЯ-настройка.pdf` в папке.

In [1]:
%pip install pandas scikit-learn

   ---------------------------------------- 0.0/10.0 MB ? eta -:--:--
   --------------------------- ------------ 6.8/10.0 MB 46.8 MB/s eta 0:00:01
   -------------------------------- ------- 8.1/10.0 MB 21.0 MB/s eta 0:00:01
   -------------------------------------- - 9.7/10.0 MB 17.1 MB/s eta 0:00:01
   ---------------------------------------- 10.0/10.0 MB 15.7 MB/s  0:00:00
   ---------------------------------------- 0.0/8.4 MB ? eta -:--:--
   ---------------- ----------------------- 3.4/8.4 MB 16.8 MB/s eta 0:00:01
   ---------------------- ----------------- 4.7/8.4 MB 11.3 MB/s eta 0:00:01
   ------------------------------ --------- 6.3/8.4 MB 9.6 MB/s eta 0:00:01
   ------------------------------------ --- 7.6/8.4 MB 8.9 MB/s eta 0:00:01
   ---------------------------------------- 8.4/8.4 MB 8.6 MB/s  0:00:00
   ---------------------------------------- 0.0/12.7 MB ? eta -:--:--
   -------- ------------------------------- 2.6/12.7 MB 13.1 MB/s eta 0:00:01
   ------------------ --

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


---
# Часть А. Журнал входов как таблица

## А1. Среда

**Ячейка А1.** Версия Python: нужна 3.11 или новее.

In [2]:
import sys
print(sys.version)

3.14.7 (tags/v3.14.7:823f032, Aug  5 2026, 10:51:32) [MSC v.1944 64 bit (AMD64)]


**Ячейка А2.** Библиотеки на месте: две версии, без ошибок.

In [3]:
import pandas as pd
import sklearn
print(pd.__version__, sklearn.__version__)

3.0.5 1.9.1


## А2. Открываем журнал

**Ячейка А3.** Загрузить файл и показать первые 5 строк.

In [4]:
df = pd.read_csv("auth_log.csv")
df.head()

,timestamp,user,ip,result
0,2026-09-01 08:09:00,kiseleva,10.0.1.38,success
1,2026-09-01 08:20:00,sidorov,10.0.1.34,failure
2,2026-09-01 08:21:00,orlova,10.0.1.25,success
3,2026-09-01 08:21:00,kozlov,10.0.1.34,success
4,2026-09-01 08:21:00,petrova,10.0.1.38,success


**Вопрос А3.** Что записано в одной строке таблицы? Какие колонки могут стать признаками? Есть ли в файле метка «атака / норма»?

**Ответ:** _(Временная метка, имя пользователя, ip, результат. В роли признака можно использовать колонку ip, временная колонка и колонка и именами пользователей, . В файле нет колонки с меткой атака/норма)_

**Ячейка А4.** Размер таблицы: (строк, колонок).

In [5]:
df.shape

(1625, 4)

**Вопрос А4.** Сколько примеров в выборке? Много это или мало для настоящего журнала организации?

**Ответ:** _(1625 строк с 4 колонками, Мне кажется что в настоящем журнале должно быть гараздо больше записей.)_

**Ячейка А5.** Тип каждой колонки. `object` или `str` — так pandas называет текст.

In [6]:
df.dtypes

timestamp    str
user         str
ip           str
result       str
dtype: object

**Вопрос А5.** Колонка `timestamp` — число или текст? Почему это важно для модели?

**Ответ:** _(текст, для модели важно что бы колонки были числовыми)_

**Ячейка А6.** Сколько раз встречается каждое значение в колонке `result`.

In [7]:
df["result"].value_counts()

result
success    898
failure    727
Name: count, dtype: int64

**Вопрос А6.** Что бросается в глаза в соотношении успехов и неудач?

**Ответ:** _(большое количество неудачных попыток входа)_

**Ячейка А7.** Доля неудачных попыток.

In [8]:
len(df[df["result"] == "failure"]) / len(df)

0.4473846153846154

**Вопрос А7.** Прочитайте выражение изнутри наружу: что делает `df["result"] == "failure"`, что делает `df[...]`, что делает `len(...)`?

**Ответ:** _(`df["result"] == "failure"` - подсчитывает количество попыток входа с неудачными исходом, `df[...]` - обращение к таблице, `len(...)` - )_

**Ячейка А8.** С каких адресов больше всего неудач (топ-10).

In [9]:
df[df["result"] == "failure"]["ip"].value_counts().head(10)

ip
45.155.205.12     259
185.220.101.34    243
91.240.118.7      164
10.0.1.34           5
10.0.1.18           5
10.0.1.39           4
10.0.1.23           4
10.0.1.24           4
10.0.1.14           4
10.0.1.35           3
Name: count, dtype: int64

**Вопрос А8.** Чем первые три адреса отличаются от остальных? Что происходит? Как назвать это в отчёте об инциденте?

**Ответ:** _(с первых трёх адресов происходит слишком много неудачных попыток входа, скорее всего происходит атака, в отчёте можно записать как "попытка атаки")_

**Ячейка А9.** Новый признак: час суток. Затем — в какие часы происходят неудачи.

In [10]:
df["hour"] = pd.to_datetime(df["timestamp"]).dt.hour
df[df["result"] == "failure"]["hour"].value_counts().sort_index()

hour
2     667
4       2
7       1
8       5
9       5
10      6
11      8
12      7
13      4
14      4
15      2
16      5
17      3
18      3
19      3
20      1
23      1
Name: count, dtype: int64

**Вопрос А9.** В какие часы неудачи? Почему это подозрительно? Назовите два признака для будущей модели, которых **не было** в файле.

**Ответ:** _(больше всего неудач в 2 часа ночи, это подозрительн тем, что очнь большое количество попыток входа в позднее время суток. ывыв)_

## А3. Глубже

**Ячейка А10.** Что происходило с учётной записью `admin`.

In [11]:
df[df["user"] == "admin"]["result"].value_counts()

result
failure    107
success      2
Name: count, dtype: int64

**Вопрос А10.** Сколько успешных входов под `admin`? Откуда и когда они были? Это попытка или уже инцидент?

Подсказка: `df[(df["user"] == "admin") & (df["result"] == "success")]` — знак `&` это «и», каждое условие в своих скобках.

**Ответ:** _(Всего успешных входов в учётную запись admin было 2, они были с ip 185.220.101.34 и 91.240.118.7 )_

In [12]:
# место для подсказки из вопроса А10
df[(df["user"] == "admin") & (df["result"] == "success")]

,timestamp,user,ip,result,hour
475,2026-09-03 02:16:42,admin,185.220.101.34,success,2
1046,2026-09-04 02:36:27,admin,91.240.118.7,success,2


**Ячейка А11.** Только внутренние адреса (сеть `10.0.1.x`).

In [13]:
df[df["ip"].str.startswith("10.0.1.")]["result"].value_counts()

result
success    888
failure     61
Name: count, dtype: int64

**Вопрос А11.** Какая доля неудач внутри сети? Какой вывод про сотрудников?

**Ответ:** _(доля неудач состваила приверно 7%.)_

**Ячейка А12.** Под какие имена пользователей подбирали пароль.

In [ ]:
df[df["result"] == "failure"]["user"].value_counts().head(10) 

user
test         118
guest        118
ivanov       115
backup       109
admin        107
root         102
zaharova       5
kozlov         5
petrova        4
kuznecova      4
Name: count, dtype: int64

**Вопрос А12.** Почему атакующий перебирал именно эти имена?

**Ответ:** _(Злоумышленники хотели получить доступ к учётной записи админа)_

---
# Часть Б. Пять задач → тип задачи машинного обучения

Четыре типа: **классификация** (на выходе категория), **регрессия** (число),
**кластеризация** (группы без названий), **поиск аномалий** («не похоже на обычное»).
Первые два — с учителем (в выборке есть правильный ответ), вторые два — без.

Заполните таблицу. Формула ответа: **тип · что на выходе · какие данные нужны · есть ли метка**.
Таблицу можно редактировать прямо здесь (два щелчка по ячейке).

| № | Задача | Тип | Что на выходе | Какие данные | Метка есть? |
|---|---|---|---|---|---|
| 1 | По тексту письма определить, фишинговое ли оно |Классификация |предупреждение о возможном фишинге |текст письма |да |
| 2 | Разбить 10 000 алертов SIEM за месяц на группы похожих, чтобы аналитик разбирал группы |кластеризация |формирование групп похожих алертов |данные из алертов: ip, время, пользователь, и т.д. |нет |
| 3 | Предсказать, через сколько дней заполнится диск сервера журналов |регрессия |вывод количества дней |объём диска, история заполнения |да |
| 4 | Сотрудник вошёл в 3 ночи с нового устройства и скачал 5 ГБ. Система должна заметить «не похоже на него» |поиск аномалий |оценка аномальности |история входов, история объёма скаченных данных  |нет |
| 5 | По описанию уязвимости оценить её опасность |классификация |уровень опасности |описание уязвимости |да |

**Вопрос Б6.** Задача «по данным DLP решить, увольнять ли сотрудника». Что с ней не так с точки зрения машинного обучения?

**Ответ:** _(не знаю)_

**Вопрос Б7 (по плану курса).** Одна задача защиты информации с вашего места практики
(или из любой знакомой организации). Опишите её той же формулой: тип, что на выходе,
какие данные нужны, есть ли метка и откуда её взять. Три–пять предложений.

**Ответ:** _(не знаю)_

---
# Часть В. Домашняя практика: меняем код под свой вопрос

Все задания решаются **изменением** ячеек из части А, писать с нуля ничего не надо.
В каждом задании подсказка, какую ячейку взять за основу. Ответ — код в ячейке
плюс одно предложение словами.

## В1. Кто из сотрудников чаще всех ошибается паролем

Только внутренние адреса (`10.0.1.x`), только неудачи, топ-5 пользователей.
За основу: ячейки А11 (фильтр по адресу) и А12 (счёт по пользователям).
Два условия соединяются через `&`, каждое в своих скобках.

In [33]:
# В1
df[(df["ip"].str.startswith("10.0.1.")) & (df["result"] == "failure")]["user"].value_counts().head(5) 

user
zaharova     5
kozlov       5
petrova      4
kuznecova    4
sokolov      4
Name: count, dtype: int64

**Вывод В1 словами:** _(zaharova 5, kozlov 5, petrova 4, kuznecova 4, sokolov 4. Сотрудники должны знать свои пароли лучше)_

## В2. Когда работал каждый из трёх внешних адресов

Для каждого из адресов `45.155.205.12`, `185.220.101.34`, `91.240.118.7` — первая и
последняя попытка по времени. Новая конструкция: `.groupby("ip")["timestamp"].agg(["min", "max"])`
считает минимум и максимум внутри каждой группы. Сначала отфильтруйте только
внешние адреса: `~df["ip"].str.startswith("10.0.1.")` — знак `~` это «не».

In [ ]:
# В2
не знаю

**Вывод В2 словами:** _(не знаю)_

## В3. Самый нагруженный день

Сколько попыток входа было в каждый день, и в какой день больше всего.
Новый признак — дата: `df["date"] = pd.to_datetime(df["timestamp"]).dt.date`
(как час в ячейке А9, только `.dt.date`). Дальше — `value_counts()`.

In [ ]:
# В3
не знаю

**Вывод В3 словами:** _(не знаю)_

## В4. Таблица признаков для будущей модели

Добавьте в `df` колонку `is_external`: 1, если адрес **не** начинается с `10.0.1.`,
иначе 0. Подсказка: `(~df["ip"].str.startswith("10.0.1.")).astype(int)`.
Затем посчитайте, сколько строк, где `is_external == 1` **и** `hour < 6`,
и сколько из них неудачных. Покажите `df.head()` — в таблице должны появиться
колонки `hour` и `is_external`.

In [ ]:
# В4
не знаю

**Вывод В4 словами:** _(не знаю)_

## В5. Сохранить результат в файл

Топ-10 адресов по неудачам (ячейка А8) сохраните в файл `top_failed_ips.csv`:
допишите к выражению `.to_csv("top_failed_ips.csv")`. Затем откройте файл
обратно: `pd.read_csv("top_failed_ips.csv")` — убедитесь, что там 10 строк.
Файл положите в репозиторий рядом с блокнотом.

In [ ]:
# В5
не знаю

## В6. Свой вопрос к данным

Придумайте свой вопрос к журналу, на который ещё не отвечали, и ответьте кодом.
Примеры: с какого внутреннего адреса больше всего успешных входов; сколько разных
пользователей заходило с каждого внешнего адреса; есть ли сотрудники, которые
входили ночью с внутреннего адреса.

**Мой вопрос:** _(не знаю)_

In [ ]:
# В6
не знаю

**Ответ словами:** _(не знаю)_

---
# Как сдавать

1. Сохраните блокнот (Ctrl + S). Убедитесь, что **все ячейки выполнены и вывод виден**:
   сверху кнопка **Run All**, потом Ctrl + S. Проверяться будет вывод, а не только код.
2. Переименуйте файл: `ДЗ-03-Фамилия.ipynb`.
3. Залейте на GitHub любым из двух способов.

**Способ 1, без установки git (через браузер).** github.com → Sign up, если нет
аккаунта → зелёная кнопка **New** (репозиторий) → имя `tech-ai-2026`, Public →
Create repository → **Add file → Upload files** → перетащите `ДЗ-03-Фамилия.ipynb`,
`auth_log.csv` и `top_failed_ips.csv` → **Commit changes**.

**Способ 2, из VS Code.** Слева значок Source Control (Ctrl + Shift + G) →
**Publish to GitHub** → Public → VS Code попросит войти в GitHub. Нужен
установленный git (git-scm.com).

4. Ссылку на репозиторий вида `https://github.com/ваш-логин/tech-ai-2026` —
   преподавателю в мессенджер группы.

Что оценивается: части А и Б — выполнены все ячейки и заполнены все ответы;
часть В — код работает, вывод соответствует вопросу, есть вывод словами.
В1–В5 обязательны, В6 — плюс балл.